### PR TITLE
feat: Return tracking in PR/DN

### PR DESCRIPTION
**Pull request by marination**
Docs: https://github.com/frappe/erpnext_documentation/pull/154

## Transaction
<details><summary>Purchase Receipt</summary>
  
  - Returned Qty fields in Items Table
    Screenshot 2020-07-31 at 4 01 17 PM

  - % Returned in parent doc
   Screenshot 2020-07-31 at 4 02 02 PM

  - Status 'Return Issued' added on original PR if 100% returned. If partially returned , priority given to 'To Bill'
   Screenshot 2020-07-31 at 4 03 48 PM

</details>

<details><summary>Delivery Note</summary>

  - Returned Qty fields in Items Table
    Screenshot 2020-07-31 at 8 14 34 PM

  - % Returned in parent doc
    Screenshot 2020-07-31 at 8 15 02 PM

  - Status 'Return Issued' added on original DN if 100% returned. If partially returned , priority given to 'To Bill'
    Screenshot 2020-07-31 at 8 15 33 PM

</details>

<details><summary>Reports Changes</summary>

- Added Returned Amount to **Delivered** and **Purchased** Items to be Billed Reports.
- **Pending Amount = Total Amount - Billed Amount  - Returned Amount**. Essentially only non-returned items should be counted as pending
- If **Pending Amount** > 0, only then will the item be visible in the Report. So if a PR is 50 % Billed and 50% Returned, it won't be visible in the Report.

Screenshot 2020-09-10 at 6 48 42 PM

</details>

ToDo:
- [x]  Return for DN
- [x]  Tests for PR/DN
- [x]  Report Fixes (Received Items to be Billed, Delivered Items to be Billed)
- [x] Patch to set qty and % 
- [x]  Test to check Correct Mapping and Billing % Updation (bill after return and bill before return)

### CODE REVIEW

**Comment by rohitwaghchaure** [comment]
During test found below issues

1. Make purchase receipt for 3 quantity, make purchase return for 1 quantity and submit it. After that do one more purchase return entry, system shows default quantity as 3. It should show default quantity as 2 because 1 quantity already has returned.

1. Make delivery note for 3 quantity, and do sales return for 1 quantity. Then make sales invoice for 2 quantity and check delivery note status. It shows the status as "To Bill", should show the status as "Completed".

Screenshot 2020-10-28 at 11 20 27 AM

**Comment by marination** [comment]
Reports bit is already ported to v12 https://github.com/frappe/erpnext/pull/23804

File: erpnext/accounts/report/received_items_to_be_billed/received_items_to_be_billed.py [hunk 1]
```diff
@@ -14,11 +14,19 @@ def execute(filters=None):
 
 def get_column():
 	return [
-		_("Purchase Receipt") + ":Link/Purchase Receipt:120", _("Status") + "::120", _("Date") + ":Date:100",
-		_("Supplier") + ":Link/Supplier:120", _("Supplier Name") + "::120",
-		_("Project") + ":Link/Project:120", _("Item Code") + ":Link/Item:120",
-		_("Amount") + ":Currency:100", _("Billed Amount") + ":Currency:100", _("Amount to Bill") + ":Currency:100",
-		_("Item Name") + "::120", _("Description") + "::120", _("Company") + ":Link/Company:120",
+		_("Purchase Receipt") + ":Link/Purchase Receipt:160",
```
Comment (gavindsouza) [inline erpnext/accounts/report/received_items_to_be_billed/received_items_to_be_billed.py]: Use JSON style API response

File: erpnext/controllers/sales_and_purchase_return.py [hunk 2]
```diff
@@ -203,10 +203,46 @@ def get_already_returned_items(doc):
...
+			sum(abs(child.stock_qty)) as stock_qty,
+			%(columns)s
+		from
+			`tab{0}` child, `tab{1}` parent
+		where
+			child.parent = parent.name
+			and parent.docstatus = 1
+			and parent.is_return = 1
+			and {2} = %(row_name)s
+		""".format(child_doctype, doctype, reference_field),
```
Comment (gavindsouza) [inline erpnext/controllers/sales_and_purchase_return.py]: ~`.format`~. Try using ORM instead.

File: erpnext/controllers/sales_and_purchase_return.py [hunk 3]
```diff
@@ -203,10 +203,46 @@ def get_already_returned_items(doc):
...
+		{
+			"row_name": row_name,
+			"columns": columns,
+			"child_doctype": child_doctype,
+			"doctype": doctype,
+			"reference_field": reference_field
+		},
+		as_dict=1)
+
+	return data[0]
```
Comment (gavindsouza) [inline erpnext/controllers/sales_and_purchase_return.py:230]: `IndexError`?

**Review by gavindsouza — CHANGES_REQUESTED** [review]

**Comment by nabinhait** [comment]
@marination resolve conflict

**Review by rohitwaghchaure — APPROVED** [review]

**Pull request merged**